<launch>
    <!-- Define arguments with default values -->
    <arg name="image_topic" default="/camera/image_raw" />
    <arg name="pose_topic" default="/orb_slam2_mono/pose" />
    <arg name="map_topic" default="/orb_slam2_mono/map" />
    <arg name="rendered_image_topic" default="/orb_slam2_mono/rendered_image" />
    <arg name="fx" default="381.36246688113556" />
    <arg name="fy" default="381.36246688113556" />
    <arg name="cx" default="320.5" />
    <arg name="cy" default="240.5" />
    <arg name="k1" default="0.0" />
    <arg name="k2" default="0.0" />
    <arg name="p1" default="0.0" />
    <arg name="p2" default="0.0" />
    <arg name="k3" default="0.0" />
    <arg name="fps" default="30" />
    <arg name="rgb_encoding" default="true" />

    <arg name="load_map" default="false" />
    <arg name="map_file_name" default="$(find orb_slam2_ros)/resource/map.bin" />

    <arg name="publish_map" default="true" />
    <arg name="publish_rendered_image" default="true" />
    <arg name="publish_pose" default="true" />
    <arg name="publish_tf" default="false" />

    <node name="orb_slam2_mono" pkg="orb_slam2_ros" type="orb_slam2_ros_mono" output="screen">
        <!-- topic remapping -->
        <remap from="/camera/image_raw" to="$(arg image_topic)" />
        <remap from="/orb_slam2_mono/pose" to="$(arg pose_topic)" />
        <remap from="/orb_slam2_mono/map" to="$(arg map_topic)" />
        <remap from="/orb_slam2_mono/rendered_image" to="$(arg rendered_image_topic)" />

        <!-- static parameters -->
        <param name="publish_map" type="bool" value="$(arg publish_map)" />
        <param name="publish_rendered_image" type="bool" value="$(arg publish_rendered_image)" />
        <param name="publish_pose" type="bool" value="$(arg publish_pose)" />
        <param name="publish_tf" type="bool" value="$(arg publish_tf)" />

        <param name="load_map" type="bool" value="$(arg load_map)" />
        <param name="map_file_name" type="string" value="$(arg map_file_name)" />

        <param name="vocabulary_file_name" type="string" value="$(find orb_slam2_ros)/orb_slam2_core/Vocabulary/ORBvoc.txt" />

        <!-- ORB parameters -->
        <param name="ORBextractor/nFeatures" type="int" value="2000" />
        <param name="ORBextractor/scaleFactor" type="double" value="1.2" />
        <param name="ORBextractor/nLevels" type="int" value="8" />
        <param name="ORBextractor/iniThFAST" type="int" value="20" />
        <param name="ORBextractor/minThFAST" type="int" value="7" />
        <!-- you don't need depth parameters for monocular case -->

        <!-- Camera parameters -->
        <param name="Camera/fx" type="double" value="$(arg fx)" />
        <param name="Camera/fy" type="double" value="$(arg fy)" />
        <param name="Camera/cx" type="double" value="$(arg cx)" />
        <param name="Camera/cy" type="double" value="$(arg cy)" />
        <param name="Camera/k1" type="double" value="$(arg k1)" />
        <param name="Camera/k2" type="double" value="$(arg k2)" />
        <param name="Camera/p1" type="double" value="$(arg p1)" />
        <param name="Camera/p2" type="double" value="$(arg p2)" />
        <param name="Camera/k3" type="double" value="$(arg k3)" />

        <param name="Camera/fps" type="int" value="$(arg fps)" />
        <param name="Camera/rgb_encoding" type="bool" value="$(arg rgb_encoding)" />
    </node>
</launch>